File: aesi-eclipse/bundles/aesi-eclipse/src/main/java/com/virtlink/aesi/structureoutline/DummyStructureOutliner.java
package com.virtlink.aesi.structureoutline;

import com.virtlink.aesi.IAesiDocument;
import com.virtlink.aesi.ICancellationToken;
import com.virtlink.aesi.Span;

import javax.annotation.Nullable;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class DummyStructureOutliner implements IStructureOutliner {

    private final SymbolDef rootSymbol = new SymbolDef("root", SymbolKind.Constant, Span.of(0, 2), Arrays.asList(
            new SymbolDef("file1", SymbolKind.File, Span.of(0, 2), Arrays.asList(
                    new SymbolDef("myClass0", SymbolKind.Class, Span.of(1, 2), Arrays.asList(
                            new SymbolDef("myFunc", SymbolKind.Function, Span.of(2, 2), Collections.emptyList())
                    )),
                    new SymbolDef("myClass1", SymbolKind.Class, Span.of(3, 2), Collections.emptyList()),
                    new SymbolDef("myClass2", SymbolKind.Class, Span.of(4, 2), Collections.emptyList()),
                    new SymbolDef("myClass4", SymbolKind.Class, Span.of(5, 2), Collections.emptyList())
            )),
            new SymbolDef("file2", SymbolKind.File, Span.of(6, 2), Arrays.asList(
                    new SymbolDef("anotherFunc", SymbolKind.Function, Span.of(7, 2), Collections.emptyList())
            ))
    ));

    @Override
    public List<? extends ISymbol> getRoots(IAesiDocument document, @Nullable ICancellationToken cancellationToken) {
        return Arrays.asList(rootSymbol);
    }

    @Override
    public List<? extends ISymbol> getChildren(IAesiDocument document, ISymbol symbol, @Nullable ICancellationToken cancellationToken) {
        return ((SymbolDef)symbol).getChildren();
    }

    @Override
    public boolean hasChildren(IAesiDocument document, ISymbol symbol, @Nullable ICancellationToken cancellationToken) {
        return getChildren(document, symbol, cancellationToken).size() > 0;
    }

    private final class SymbolDef extends Symbol {
        private final List<SymbolDef> children;

        public List<SymbolDef> getChildren() { return this.children; }

        public SymbolDef(String name, SymbolKind kind, Span location, List<SymbolDef> children) {
            super(name, kind, location);

            this.children = children;
        }
    }
}

File: paplj/src/main/kotlin/com/virtlink/paplj/structureoutline/DummyStructureOutliner.kt
package com.virtlink.paplj.structureoutline

import com.virtlink.editorservices.ICancellationToken
import com.virtlink.editorservices.ScopeNames
import com.virtlink.editorservices.Span
import com.virtlink.editorservices.structureoutline.*
import com.virtlink.editorservices.symbols.ISymbol
import com.virtlink.editorservices.symbols.Symbol
import com.virtlink.logging.logger
import java.net.URI

class DummyStructureOutliner: IStructureOutlineService {

    @Suppress("PrivatePropertyName")
    private val LOG by logger()

    override fun configure(configuration: IStructureOutlineConfiguration) {
        // Nothing to do.
    }

    override fun getRoots(document: URI, cancellationToken: ICancellationToken?): IStructureOutlineInfo
            = StructureOutlineInfo(listOf(rootSymbol))

    override fun getChildren(document: URI, node: IStructureOutlineElement, cancellationToken: ICancellationToken?): IStructureOutlineInfo
            = StructureOutlineInfo((node as Node).children)

    private val rootSymbol: Node = Node(label = "root", scopes = ScopeNames("meta.constant"), children = listOf(
            Node(label = "file1", scopes = ScopeNames("meta.file"), children = listOf(
                    Node(label = "myClass0", scopes = ScopeNames("meta.class"), children = listOf(
                            Node(label = "myFunc", scopes = ScopeNames("meta.function"), children = listOf())
                    )),
                    Node(label = "myClass2", scopes = ScopeNames("meta.class"), children = listOf()),
                    Node(label = "myClass4", scopes = ScopeNames("meta.class"), children = listOf())
            )),
            Node(label = "file2", scopes = ScopeNames("meta.file"), children = listOf())
    ))

    private data class Node(
            val children: List<Node>,
            override val label: String,
            override val nameSpan: Span? = null,
            override val scopes: ScopeNames = ScopeNames(),
            override val isLeaf: Boolean? = null)
        : IStructureOutlineElement

}